File: Python-Src/network.py
from math import exp, pow
from pipe import Pipe
from state import State
from node import Node

class Network:

    def __init__(self, learn_rate = 0.001, layers = []):
        self.learn_rate = learn_rate
        self.layers = layers
        self.input_layer = []
        self.output_layer = []

    # THIS IS "ABSTRACT METHOD"
    def build_network(self):
        pass

    # THIS IS "ABSTRACT METHOD"
    def train_network(self, iterations=0, cases=[]):
        pass

    def connect_pipeline(self, source_node, target_node):
        pipe_line = Pipe()
        source_node.add_front_pipe(pipe_line)
        target_node.add_back_pipe(pipe_line)

    def connect_state_to_node(self, source_state, target_node):
        if source_state.pipe_line == None:
            pipe_line = Pipe()
            source_state.add_pipeline(pipe_line)
            target_node.add_back_pipe(pipe_line)
        else:
            target_node.add_back_pipe(source_state.pipe_line)

    def connect_node_to_state(self, source_node, target_state):
        if target_state.pipe_line == None:
            pipe_line = Pipe()
            source_node.add_front_pipe(pipe_line)
            target_state.add_pipeline(pipe_line)
        else:
            source_node.add_front_pipe(target_state.pipe_line)

    def add_layer(self, layer):
        self.layers.append(layer)

    def run_normal(self, inputs, return_raw=False):
        self.__forward(inputs)
        ans = []
        for output_state in self.output_layer:
            if return_raw:
                ans.append(output_state.state_value)
            else:
                if (0.5 <= output_state.state_value):
                    ans.append(1)
                else:
                    ans.append(0)
        return ans

    def run_forward(self, inputs):
        self.__forward(inputs)

    def __forward(self, inputs):
        if len(inputs) != len(self.input_layer):
            print("Inputs don\'t aline with input states. Please check again.")
        for index in range(0, len(inputs)):
            self.input_layer[index].load_value(float(inputs[index]))
        for input_state in self.input_layer:
            input_state.pipe_line.forward_result = input_state.state_value
        for layer in self.layers:
            layer.forward()
        for output_state in self.output_layer:
            output_state.load_value(output_state.pipe_line.forward_result)

    def run_backward(self, expected):
        self.__backward(expected)

    def __backward(self, expected):
        if len(expected) != len(self.output_layer):
            print("Error: expected and actual length isn\'t valid.")
            return
        constant = 1.0 / len(expected)
        res = []
        for index in range(0, len(expected)):
            self.output_layer[index].pipe_line.backward_result = (-1.0 * constant * (expected[index] - self.output_layer[index].state_value))
        index_layer = len(self.layers) - 1
        while index_layer >= 0:
            # print(self.layers[index_layer].to_string())
            self.layers[index_layer].backward(self.learn_rate)
            index_layer -= 1

File: Swift-Src/Swift-Src/network.swift
//
//  network.swift
//  Swift-Src
//
//  Created by Wesley Webb on 9/23/20.
//

import Foundation

// Figure out ambstraction in Swift
class Network
{
    private var _learningRate:Double;
    var learningRate:Double {
        get { return self._learningRate; }
        set { self._learningRate = newValue; }
    }
    
    private var _layers:[Layer];
    var layers:[Layer] {
        get { return self._layers; }
        set { self._layers = newValue; }
    }
    
    private var _inputLayer:[State]?;
    var inputLayer:[State]? {
        get { return self._inputLayer; }
        set { self._inputLayer = newValue; }
    }
    
    private var _outputLayer:[State]?;
    var outputLayer:[State]? {
        get { return self._outputLayer; }
        set { self._outputLayer = newValue; }
    }
    
    init(newLayers:[Layer] = [Layer](), newLearningRate:Double = 0.01)
    {
        self._learningRate = newLearningRate;
        self._layers = newLayers;
        self._inputLayer = nil;
        self._outputLayer = nil;
    }
    
    func buildNetwork()
    {
        print("asdf");
    }
    
    func trainNetwork(iterations:Int, cases:[[Double]])
    {
        print("fdsa");
    }
    
    func connnectPipeline(sourceNode:Node, targetNode:Node)
    {
        let newPipe:Pipe = Pipe();
        sourceNode.addFrontPipe(pipe: newPipe);
        targetNode.addBackPipe(pipe: newPipe);
    }
    
    func connectStateToNode(sourceState:State, targetNode:Node)
    {
        if (sourceState.pipeline == nil)
        {
            let pipeline:Pipe = Pipe();
            sourceState.addPipeLine(pipeLine: pipeline);
            targetNode.addBackPipe(pipe: pipeline);
        }
        else
        {
            targetNode.addBackPipe(pipe: sourceState.pipeline!);
        }
    }
    
    func connectNodeToState(sourceNode:Node, targetState:State)
    {
        if (targetState.pipeline == nil)
        {
            let pipeline:Pipe = Pipe();
            sourceNode.addFrontPipe(pipe: pipeline);
            targetState.addPipeLine(pipeLine: pipeline);
        }
        else
        {
            sourceNode.addFrontPipe(pipe: targetState.pipeline!);
        }
    }
    
    func runNormal(newInputs:[Double], returnRaw:Bool = false) -> [Double]
    {
        self.foward(newInputs: newInputs);
        var ans:[Double] = [Double]();
        
        for outputState in self.outputLayer!
        {
            if (returnRaw)
            {
                ans.append(outputState.stateValue);
            }
            else
            {
                ans.append(0.5 <= outputState.stateValue ? 1.0 : 0.0);
            }
        }
        
        return ans;
    }
    
    func runFoward(newInputs:[Double])
    {
        self.foward(newInputs: newInputs)
    }
    
    private func foward(newInputs:[Double])
    {
        for index in 0..<newInputs.count
        {
            self.inputLayer![index].loadValue(stateValue: newInputs[index]);
        }
        for state in self.inputLayer!
        {
            state.pipeline?.forwardResult = state.stateValue;
        }
        for layer in self.layers
        {
            layer.forward();
        }
        for state in self.outputLayer!
        {
            state.loadValue(stateValue: state.pipeline!.forwardResult)
        }
    }
    
    func runBackward(newExpected:[Double])
    {
        self.backward(newExpected: newExpected);
    }
    
    private func backward(newExpected:[Double])
    {
        let constant:Double = 1.0 / Double(newExpected.count);
        for index in 0..<newExpected.count
        {
            let paramValue:Double = (-1.0 * constant * (newExpected[index] - self.outputLayer![index].stateValue));
            self.outputLayer?[index].pipeline?.backwardResult = paramValue;
        }
        
        var indexLayer:Int = self.layers.count - 1;
        while (indexLayer >= 0)
        {
            self.layers[indexLayer].backward();
            indexLayer -= 1;
        }
    }
}
